<a href="https://colab.research.google.com/github/Renato-Rsprojetos3D/Machine-Learning-Data-Sciences/blob/master/Aula_01_Quarentena_Alura.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![alt text](https://www.alura.com.br/assets/img/alura-share.1571848411.png)

# Importando as Bibliotecas necessarias

In [0]:
import pandas as pd
import numpy as np

# Carregando os arquivos CSV "movies" e "ratings"

In [0]:


# Carregando ratings.csv
avaliacoes = pd.read_csv('/content/drive/My Drive/ratings.csv')

# Alterando os nomes das colunas do dataframe avaliacoes
avaliacoes.columns = ['usuarioId','filmeId','notas','momento']

# Carregando ratings.csv
filmes = pd.read_csv('/content/drive/My Drive/movies.csv')

# Alterando os nomes das colunas do dataframe filmes
filmes.columns= ['filmeId','titulo','generos']

In [3]:
# Visualizando as 5 primeiras linhas do dataframe avaliacoes
avaliacoes.head()

usuarioId  filmeId  notas    momento
0          1        1    4.0  964982703
1          1        3    4.0  964981247
2          1        6    4.0  964982224
3          1       47    5.0  964983815
4          1       50    5.0  964982931

In [4]:
# Visualizando numeros de linhas e colunas dataframe avalicoes
avaliacoes.shape

(100836, 4)

In [5]:
# Visualizando as 5 primeiras linhas do dataframe filmes
filmes.head()

filmeId  ...                                      generos
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

In [6]:
# Visualizando numeros de linhas e colunas dataframe filmes
filmes.shape

(9742, 3)

![alt text](https://minerandodados.com.br/wp-content/uploads/2019/12/pandas-960x720.png)

# Obtendo as médias dos filmes no dataframe  avaliações

In [7]:
# Agrupando por " filmeId " e obtendo a média das notas dos filmes
notas_medias_por_filme = avaliacoes.groupby('filmeId')['notas'].mean()

# Visualizando as 5 primeiras linhas
notas_medias_por_filme.head()

filmeId
1    3.920930
2    3.431818
3    3.259615
4    2.357143
5    3.071429
Name: notas, dtype: float64

![alt text](https://data-flair.training/blogs/wp-content/uploads/sites/2/2019/06/Pandas-Merge-and-Join-Functions-1280x720.jpg)

# juntando a colunas nota com as medias de avaliação ao dataframe filmes

In [8]:
# juntando a colunas nota com as medias de avaliação ao dataframe filmes
filmes_com_media = filmes.join(notas_medias_por_filme,on= 'filmeId')

# Visualizando as 5 primeiras linhas do dataframe
filmes_com_media

filmeId  ...     notas
0           1  ...  3.920930
1           2  ...  3.431818
2           3  ...  3.259615
3           4  ...  2.357143
4           5  ...  3.071429
...       ...  ...       ...
9737   193581  ...  4.000000
9738   193583  ...  3.500000
9739   193585  ...  3.500000
9740   193587  ...  3.500000
9741   193609  ...  4.000000

[9742 rows x 4 columns]

![alt text](https://media.giphy.com/media/ugFzEiuHAvYnm/giphy.gif)

# Desafio 1

Encontre os 18 filmes que não tiveram avaliação

In [9]:
# Verificando se dataframe possui valores faltantes
filmes_com_media.isna().sum()

filmeId     0
titulo      0
generos     0
notas      18
dtype: int64

# Conclusões do desafio 1 e 2
Após verificar se havia dados ausentes foi encontrado na coluna " notas " que existe 18 valores ausentes , respondendo o desafio 1 , isso pode ocorrer pela falta de preenchimento na hora de dar a nota ao filme.

No desafio 2 ele foi realizado logo no inicio do Notebook onde renomei todas as colunas

![alt text](https://media.giphy.com/media/6mlX32zHKWKTSOj2mP/giphy.gif)

# Desafio 3

Colocar o número de avaliações por filme , isto é , não so a média mais o total de votos por filme

In [20]:
# Agrupando por "filmeId , titulo, generos" e contando numero de votos
desafio_3 = filmes_com_media.groupby(['filmeId','titulo','generos'])['notas'].count()
desafio_3

filmeId  titulo                                     generos                                    
1        Toy Story (1995)                           Adventure|Animation|Children|Comedy|Fantasy    1
2        Jumanji (1995)                             Adventure|Children|Fantasy                     1
3        Grumpier Old Men (1995)                    Comedy|Romance                                 1
4        Waiting to Exhale (1995)                   Comedy|Drama|Romance                           1
5        Father of the Bride Part II (1995)         Comedy                                         1
                                                                                                  ..
193581   Black Butler: Book of the Atlantic (2017)  Action|Animation|Comedy|Fantasy                1
193583   No Game No Life: Zero (2017)               Animation|Comedy|Fantasy                       1
193585   Flint (2017)                               Drama                                       

# Conclusões do desafio 3


![alt text](https://media.giphy.com/media/MDQwpbdpOdVrW/giphy.gif)

# Desafio 4

Arredondar os valores da coluna " notas " para e casas decimais 

In [11]:
# Arredondando a coluna " notas " para 2 casas decimais
filmes_com_media['notas'] = filmes_com_media.notas.round(2)

# Visulizando os 5 primeiros registros
filmes_com_media.head()

filmeId  ... notas
0        1  ...  3.92
1        2  ...  3.43
2        3  ...  3.26
3        4  ...  2.36
4        5  ...  3.07

[5 rows x 4 columns]

![alt text](https://media.giphy.com/media/xTiIzAWLqreqgY1I0o/giphy.gif)

# Desafio 5
Descobrir os Generos dos filmes(quais são eles, unicos)



In [25]:
desafio_5 = filmes_com_media['generos'].str.get_dummies('|').keys()
desafio_5

Index(['(no genres listed)', 'Action', 'Adventure', 'Animation', 'Children',
       'Comedy', 'Crime', 'Documentary', 'Drama', 'Fantasy', 'Film-Noir',
       'Horror', 'IMAX', 'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Thriller',
       'War', 'Western'],
      dtype='object')

![alt text](https://media.giphy.com/media/x1XqoA5TxnrfW/giphy.gif)

# Desafio 6

Contar o numero de aparações de cada genero

In [26]:
desafio_6 = filmes_com_media['generos'].str.get_dummies('|').sum().sort_values(ascending = False)
desafio_6

Drama                 4361
Comedy                3756
Thriller              1894
Action                1828
Romance               1596
Adventure             1263
Crime                 1199
Sci-Fi                 980
Horror                 978
Fantasy                779
Children               664
Animation              611
Mystery                573
Documentary            440
War                    382
Musical                334
Western                167
IMAX                   158
Film-Noir               87
(no genres listed)      34
dtype: int64

# Conclusões do desafio 6

Notamos que genero " Drama " é avaliado 4361 vezes

![alt text](https://media.giphy.com/media/8SxGru3XzElqg/giphy.gif)

# Desafio 7

Criar um grafico de aparições por Genero



In [0]:
# importando a biblioteca mateplotlib para criar graficos 
import matplotlib.pyplot as plt

In [0]:
# Ordenando os valores de maior para menor
valores = filmes_com_media['generos'].str.get_dummies('|').sum().sort_values(ascending = False)

In [35]:
from plotly.offline import plot, iplot, init_notebook_mode
init_notebook_mode(connected=True)
import plotly.graph_objs as go
# Definindo o renderizador:
import plotly.io as pio
pio.renderers
pio.renderers.default = "colab"

value = valores
label = value.index
trace = go.Bar(y=value, 
               x=label,text = value,textposition= "outside",
               marker={'color': value,
                   'colorscale': 'PiYG'})
layout = go.Layout(title="Generos por Aparições", 
                   xaxis={'title' : "Generos"},
                  yaxis={'title' : "Aparições"},
                  title_x=0.5)
fig = go.Figure(data=trace, layout=layout)
iplot(fig)

# Conclusões do desafio 7

No grafico acima podemos notar que o genero " Drama " seguido por "Comédia " possuim o maior numero de aparições no dataframe